### DISCUSSION TITLE
Use of ActionToken without Authentication Session for user activation will result in cookie errors

### DISCUSSION THREAD

**Discussion by twobiers** [keycloak/keycloak#29257] [Q&A]
Hello Keycloak Community,

we're currently relying on a customized user activation process where a user registration will be performed by a backend call and sends out an activation email. Therefore we don't have an active authentication session for the user and it is also not present in the token. Every newly registered user has an required action to set an initial password for itself. Therfore will be redirected to the required action during action token handling.

A problem comes into play when the user uses the action token in two separate tabs. Because the LoginActionsService will perform a fresh login everytime the action token is used, the authentication state is only valid for the most recent opened tab. When the password is being set in one of the previous tabs, it will result in a `cookie_not_found` `CUSTOM_REQUIRED_ACTION_ERROR`, because the new cookie is not associated with the Tab ID.
We see that this happens regular in our systems.

Currently we plan to introduce an intermediate verification step for fresh authentication sessions like it is in the default VerifyEmail action token handler, but is there anything else that can be done to mitigate this? Is it maybe something which should be covered in the authChecker script?

**Comment by kmharish** [comment]
Were you able to figure this out, I am trying to do something similer.

**Reply by twobiers** [reply]
Hi, inside of the TokenHandler `handleToken()` method, we have put the following snippet:

```java
if (tokenContext.isAuthenticationSessionFresh()) {
    String authSessionEncodedId = AuthenticationSessionCompoundId.fromAuthSession(authSession).getEncodedId();
     token.setCompoundAuthenticationSessionId(authSessionEncodedId);
     UriBuilder builder = Urls.actionTokenBuilder(uriInfo.getBaseUri(), token.serialize(session, realm, uriInfo),
     authSession.getClient().getClientId(), authSession.getTabId(), AuthenticationProcessor.getClientData(session, authSession));
     String confirmUri = builder.build(realm.getName()).toString();

      return session.getProvider(LoginFormsProvider.class)
         .setAuthenticationSession(authSession)
         .setSuccess(Messages.CONFIRM_EMAIL_ADDRESS_VERIFICATION, user.getEmail())
         .setAttribute(Constants.TEMPLATE_ATTR_ACTION_URI, confirmUri)
         .createInfoPage();
 }
```

This forces the user to perform an additional redirect and allows the cookies to be set accordingly.
